### PR TITLE
expression, executor: rewrite builtin func reverse.

### PR DESCRIPTION
**Pull request by yangwenmai**
expression, executor: rewrite builtin func reverse.

### CODE REVIEW

File: expression/builtin_string.go [hunk 1]
```diff
@@ -453,38 +453,45 @@ func (b *builtinLowerSig) eval(row []types.Datum) (d types.Datum, err error) {
...
+func oneArgsInferType(args []Expression) *types.FieldType {
+	argTp := args[0].GetType()
+	tp := types.MergeFieldType(mysql.TypeVarString, argTp.Tp)
+	retType := types.NewFieldType(tp)
+	retType.Charset, retType.Collate = charset.CharsetUTF8, charset.CollationUTF8
+	if types.IsBinaryStr(argTp) {
+		retType.Charset, retType.Collate = charset.CharsetBin, charset.CollationBin
+		retType.Flag |= mysql.BinaryFlag
+	}
+	return retType
```
Comment (zz-jason) [inline expression/builtin_string.go]: please also set `retType`'s `Flen` before return or during the evaluation of `builtinReverseSig`

**Comment by XuHuaiyu** [comment]
@yangwenmai 
we've update newBaseBuiltinFuncWithTp(),
please update this PR.

for a function which returns string,
we should set the flen, chs, collate and flag to correct value.

**Comment by XuHuaiyu** [comment]
ci failed @yangwenmai

**Comment by yangwenmai** [comment]
I rewrite #3697 .

**Pull request closed without merging**